File: src/console/demo_channel.py
"""Decide whether the UI should use the console channel or JWT HTTP APIs.

Demo Sign In has no bearer token. Calling ``/api/notifications`` produced
401 toasts. The console channel serves the same inbox locally.
"""

from __future__ import annotations

from src.auth.policy import is_demo_or_guest, requires_bearer_token
from src.auth.session import Session
from src.integration.authz import transport_hints
from src.integration.commands import JWT_NOTIFICATION_PATHS
from src.notifications.policy import should_fetch, should_open_stream


def should_use_console_inbox(session: Session | None, *, bearer_token: str | None = None) -> bool:
    if bearer_token or (session and session.has_bearer_token):
        return False
    return session is not None and is_demo_or_guest(session)


def should_call_jwt_path(
    path: str,
    session: Session | None,
    *,
    bearer_token: str | None = None,
) -> bool:
    normalized = str(path or "").split("?", 1)[0]
    if normalized.rstrip("/") in {item.rstrip("/") for item in JWT_NOTIFICATION_PATHS} or normalized.startswith(
        "/api/notifications/"
    ):
        if normalized.endswith("/stream") or normalized.endswith("/stream/"):
            return should_open_stream(session, bearer_token=bearer_token)
        return should_fetch(session, bearer_token=bearer_token)
    return True


def policy_payload(session: Session | None, *, bearer_token: str | None = None) -> dict[str, object]:
    hints = transport_hints(session, bearer_token=bearer_token)
    jwt_ok = requires_bearer_token(session) or bool(bearer_token)
    return {
        **hints,
        "fetch_notifications_http": jwt_ok,
        "open_notification_stream": jwt_ok,
        "use_console_inbox": should_use_console_inbox(session, bearer_token=bearer_token),
        "reason": "demo_or_guest" if is_demo_or_guest(session) else ("bearer" if jwt_ok else "anonymous"),
    }

File: frontend/src/features/bridge/demoChannel.ts
import { JWT_NOTIFICATION_PATHS } from './commands';
import { shouldCallJwtNotifications, shouldUseConsoleInbox, type GateInput } from './authGate';
import type { TransportHints } from './types';

export function transportHints(input: GateInput): TransportHints {
  const jwtOk = shouldCallJwtNotifications(input);
  const demo = Boolean(input.session && (input.session.kind === 'demo' || input.session.kind === 'guest'));
  return {
    use_console_channel: true,
    skip_jwt_notifications: !jwtOk,
    skip_jwt_notification_stream: !jwtOk,
    demo_or_guest: demo,
    has_bearer_token: jwtOk,
    fetch_notifications_http: jwtOk,
    open_notification_stream: jwtOk,
    use_console_inbox: shouldUseConsoleInbox(input),
    reason: demo ? 'demo_or_guest' : jwtOk ? 'bearer' : 'anonymous',
  };
}

export function shouldFetchPath(path: string, input: GateInput): boolean {
  const normalized = path.split('?')[0] ?? path;
  const jwtInbox = JWT_NOTIFICATION_PATHS.some(
    (prefix) => normalized === prefix || normalized.startsWith('/api/notifications/'),
  );
  if (jwtInbox) {
    return shouldCallJwtNotifications(input);
  }
  return true;
}

/**
 * Wrap existing notification hooks: skip the JWT URL after Demo Sign In
 * so the UI does not spam 401.
 */
export function notificationsFetchUrl(input: GateInput): string | null {
  if (shouldCallJwtNotifications(input)) {
    return '/api/notifications?limit=100&offset=0';
  }
  if (shouldUseConsoleInbox(input)) {
    return '/api/console/notifications';
  }
  return null;
}

export function notificationsStreamUrl(input: GateInput): string | null {
  if (shouldCallJwtNotifications(input) && input.bearerToken) {
    return `/api/notifications/stream?token=${encodeURIComponent(input.bearerToken)}`;
  }
  if (shouldUseConsoleInbox(input)) {
    return '/api/console/stream';
  }
  return null;
}

export function inboxWriteBase(input: GateInput): string | null {
  if (shouldCallJwtNotifications(input)) {
    return '/api/notifications';
  }
  if (shouldUseConsoleInbox(input)) {
    return '/api/console/notifications';
  }
  return null;
}

export function unwrapEnvelope<T extends Record<string, unknown>>(json: unknown): T {
  if (!json || typeof json !== 'object') {
    return {} as T;
  }
  const record = json as Record<string, unknown>;
  if (record.notifications || record.unread_count !== undefined) {
    return record as T;
  }
  if (record.data && typeof record.data === 'object') {
    return record.data as T;
  }
  return record as T;
}
